edb/schema/pointers.py

```diff
@@ -917,54 +1044,66 @@ def _process_create_or_alter_ast(self, schema, astnode, context):
         elif target_ref is not None:
             self._set_pointer_type(schema, astnode, context, target_ref)
 
-    @classmethod
-    def _extract_union_operands(cls, expr, operands):
-        if expr.op == 'UNION':
-            cls._extract_union_operands(expr.op_larg, operands)
-            cls._extract_union_operands(expr.op_rarg, operands)
-        else:
-            operands.append(expr)
-
-    def compile_expr_field(self, schema, context, field, value):
+    def compile_expr_field(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+        field: so.Field[Any],
+        value: Any,
+    ) -> s_expr.Expression:
         from . import sources as s_sources
 
         if field.name in {'default', 'expr'}:
-            singletons = []
+            singletons: List[s_types.Type] = []
             path_prefix_anchor = None
-            anchors = {}
+            anchors: Dict[str, Any] = {}
 
             if field.name == 'expr':
+                # type ignore below, because the class is used as mixin
                 parent_ctx = context.get_ancestor(
-                    s_sources.SourceCommandContext, self)
+                    s_sources.SourceCommandContext,  # type: ignore
+                    self
+                )
+                assert parent_ctx is not None
                 source_name = parent_ctx.op.classname
                 source = schema.get(source_name, default=None)
                 anchors[qlast.Source().name] = source
                 if not isinstance(source, Pointer):
+                    assert source is not None
                     singletons = [source]
                     path_prefix_anchor = qlast.Source().name
 
-            return type(value).compiled(
+            expression = type(value).compiled(
                 value,
                 schema=schema,
                 options=qlcompiler.CompilerOptions(
                     modaliases=context.modaliases,
                     schema_object_context=self.get_schema_metaclass(),
                     anchors=anchors,
                     path_prefix_anchor=path_prefix_anchor,
-                    singletons=singletons,
+                    singletons=frozenset(singletons),
                 ),
             )
+            assert isinstance(expression, s_expr.Expression)
```

Comment:
Why is this `assert` necessary?

Reply:
I did a mistake here: `value` was erroneously annotated as 'Any'. I corrected by saying that `value: s_expr.Expression`.

Reply:
I probably copied the annotations from the parent class in `delta.ObjectCommand`.